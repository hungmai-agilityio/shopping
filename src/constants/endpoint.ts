export const END_POINT = {
  HOME: '/',
  SHOP: '/shop',
  ABOUT_US: '/about',
  CONTACT: '/contact',
  SIGN_IN: '/signIn',
  SIGN_UP: '/signUp',
  WISHLIST: '/wishlist',
  PRODUCTS: '/products',
  CATEGORY: '/category',
  COMMENT: '/comment',
  USERS: '/users',
  PROFILE: '/profiles',
  CART: '/cart',
  CHECKOUT: '/checkout'
};
