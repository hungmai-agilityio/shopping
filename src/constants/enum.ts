export enum TYPE {
  PRIMARY = 'primary',
  SECOND = 'second',
  THIRD = 'third',
  FILE = 'file',
  TEXT = 'text',
  PASSWORD = 'password',
  EMAIL = 'email',
  COMMENT = 'comment',
  CONTACT = 'contact'
}

export enum SIZE {
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large'
}

export enum TAG {
  H1 = 'h1',
  H2 = 'h2',
  H3 = 'h3',
  H4 = 'h4',
  H5 = 'h5',
  H6 = 'h6',
  P = 'p',
  SPAN = 'span'
}

export enum FONT_SIZE {
  LARGE = 'text-heading',
  MEDIUM = 'text-4xl',
  SMALL = 'text-2xl',
  X_SMALL = 'lg:text-lg text-sm',
  MIMI = 'text-sm'
}

export enum FONT_WEIGHT {
  THIN = 'thin',
  LIGHT = 'light',
  NORMAL = 'normal',
  MEDIUM = 'medium',
  SEMIBOLD = 'semibold',
  BOLD = 'bold'
}

export enum CELL {
  TH = 'th',
  TD = 'td'
}

export enum CURRENT {
  USER = 'user',
  PROFILE = 'profile'
}

export enum GENDER {
  MALE = '0',
  FEMALE = '1',
  OTHER = '2'
}

export enum KEY {
  USER = 'currentUser'
}

export enum STATUS {
  SUCCESS = 'success',
  ERROR = 'error',
  WARNING = 'warning'
}

export enum TIME_STATUS {
  STALE_TIME = 1000 * 60 * 60, // milliseconds * seconds * minutes = 1 hour
  COOKIE_TIME = 60 * 60 * 24 * 2 // seconds * minutes * hour * day
}