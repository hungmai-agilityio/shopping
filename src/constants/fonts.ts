import { Mada, Poppins, Roboto, Signika } from 'next/font/google';

export const mada = Mada({
  weight: ['400', '700'],
  subsets: ['latin']
});

export const popping = Poppins({
  subsets: ['latin'],
  weight: ['400', '700']
})

export const roboto = Roboto({
  subsets: ['latin'],
  weight: ['400', '700']
})

export const signika = Signika({
  subsets: ['latin'],
  weight: ['400', '700']
})