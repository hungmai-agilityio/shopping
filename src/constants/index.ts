export * from '@/constants/fonts';
export * from '@/constants/enum';
export * from '@/constants/endpoint';
export * from '@/constants/messages';
export * from '@/constants/list';
