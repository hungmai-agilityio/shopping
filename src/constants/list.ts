import { END_POINT } from "@/constants/endpoint";

export const VALID_IMAGE = [
  'image/png',
  'image/jpeg',
  'image/jpg',
];

export const slideProducts = [
  {
    id: 'i7g6i9i2-8f89-6h15-9i0e-6j2f7j33g6i6',
    name: 'Basic T-Shirt',
    price: 15,
    description: 'A comfortable and stylish basic t-shirt.',
    category: 't-shirts',
    colors: ['White', 'Blue', 'Green'],
    sizes: ['S', 'M', 'L'],
    image: 'https://i.ibb.co/njkCQHq/t-shirt-3.webp',
    imageDetail: [
      'https://i.ibb.co/njkCQHq/t-shirt-3.webp',
      'https://i.ibb.co/njkCQHq/t-shirt-3.webp',
      'https://i.ibb.co/njkCQHq/t-shirt-3.webp'
    ],
    rating: '10k',
    sell: 'Best quality',
    stoct: 1710
  },
  {
    id: 'u3k8h9f1-9g22-3h67-2g13-5k6g9k45g4t3',
    name: 'Sporty T-Shirt',
    price: 25,
    description: 'Perfect for your workout sessions.',
    category: 't-shirts',
    colors: ['Black', 'Gray', 'white', 'Maroon'],
    sizes: ['M', 'L', 'XL'],
    image: 'https://i.ibb.co/WtgcmWR/t-shirt-2.webp',
    imageDetail: [
      'https://i.ibb.co/WtgcmWR/t-shirt-2.webp',
      'https://i.ibb.co/WtgcmWR/t-shirt-2.webp',
      'https://i.ibb.co/WtgcmWR/t-shirt-2.webp'
    ],
    rating: '1k',
    sell: 'Best quality',
    stoct: 1118
  },
  {
    id: 'y4l7i8o1-2j33-4k56-8p92-7m9b8j45h7u8',
    name: 'Casual T-Shirt',
    price: 20,
    description: 'Stay casual and comfortable.',
    category: 't-shirts',
    colors: ['White', 'Yellow', 'Pink'],
    sizes: ['XS', 'S', 'M', 'L'],
    image: 'https://i.ibb.co/J3GnxxJ/t-shirt-4.webp',
    imageDetail: [
      'https://i.ibb.co/J3GnxxJ/t-shirt-4.webp',
      'https://i.ibb.co/J3GnxxJ/t-shirt-4.webp',
      'https://i.ibb.co/J3GnxxJ/t-shirt-4.webp'
    ],
    rating: '2k',
    sell: 'Best quality',
    stoct: 1710
  },
  {
    id: 'o9b5l2e3-5t78-2v45-7h92-4c8h7d99a3u1',
    name: 'Graphic T-Shirt',
    price: 24,
    description: 'Unique graphic design for a standout look.',
    category: 't-shirts',
    colors: ['Blue', 'Purple'],
    sizes: ['S', 'M', 'L', 'XL'],
    image: 'https://i.ibb.co/fvhz0L9/t-shirt-1.webp',
    imageDetail: [
      'https://i.ibb.co/fvhz0L9/t-shirt-1.webp',
      'https://i.ibb.co/fvhz0L9/t-shirt-1.webp'
    ],
    rating: '9k',
    sell: 'Best quality',
    stoct: 1710
  }
];

export const slideImage = [
  'https://i.ibb.co/MRTgRYj/shirt-1.webp',
  'https://i.ibb.co/h1pRWGY/jacket-5-2.webp',
  'https://i.ibb.co/3d7b9SB/jean-2.webp'
];

export const navItems = [
  { name: 'Home', url: END_POINT.HOME },
  { name: 'About Us', url: END_POINT.ABOUT_US },
  { name: 'Contact', url: END_POINT.CONTACT }
];

export const itemsHome = [
  { name: 'Product', url: END_POINT.HOME },
  { name: 'Categories', url: END_POINT.HOME },
  { name: 'Login', url: END_POINT.SIGN_IN }
];

export const itemsShop = [
  { name: 'T-Shirt', url: END_POINT.HOME },
  { name: 'Jacket', url: END_POINT.HOME },
  { name: 'Shirt', url: END_POINT.HOME },
  { name: 'Jeans', url: END_POINT.HOME }
];

export const itemsCategories = [
  { name: 'Men', url: END_POINT.HOME },
  { name: 'Children', url: END_POINT.HOME },
  { name: 'Woman', url: END_POINT.HOME }
];

export const itemsContact = [{ name: 'mangcoding@gmail.com', url: '' }];