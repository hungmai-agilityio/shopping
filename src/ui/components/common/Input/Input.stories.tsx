import { Meta, StoryObj } from '@storybook/react';
import { useState } from 'react';

// Components
import Input from '@/ui/components/common/Input';

// Constants
import { TYPE } from '@/constants';

const meta: Meta<typeof Input> = {
  component: Input,
  tags: ['autodocs'],
  argTypes: {
    label: {
      description: 'Label for the input field',
      control: { type: 'text' }
    },
    placeholder: {
      description: 'Placeholder text inside the input field',
      control: { type: 'text' }
    },
    value: {
      description: 'Value of the input field',
      control: { type: 'text' }
    },
    variant: {
      description: 'Style variant of the input field',
      options: [TYPE.PRIMARY, TYPE.SECOND, TYPE.THIRD],
      control: { type: 'radio' }
    },
    type: {
      description: 'HTML type of the input field',
      options: ['text', 'password', 'email', 'number'],
      control: { type: 'select' }
    },
    icon: {
      description: 'Optional icon to display on the input field',
      control: { type: 'text' }
    },
    message: {
      description: 'Error or helper message below the input',
      control: { type: 'text' }
    },
    disabled: {
      description: 'Whether the input is disabled',
      control: { type: 'boolean' }
    },
    onClick: {
      description: 'Event handler for when the icon is clicked'
    },
    rows: {
      description: 'Number of rows for a textarea',
      control: { type: 'number' }
    }
  }
};

export default meta;

type Story = StoryObj<typeof Input>;

export const Default: Story = {
  render: () => {
    const [textInput, setTextInput] = useState('');

    const handleTextInputChange = (
      event: React.ChangeEvent<HTMLInputElement>
    ) => {
      setTextInput(event.target.value);
    };

    return (
      <Input
        label="Text Input"
        placeholder="Enter some text"
        value={textInput}
        onChange={handleTextInputChange}
        variant={TYPE.PRIMARY}
        message="This field is required"
      />
    );
  }
};

export const WithMessage: Story = {
  render: () => {
    const [email, setEmail] = useState('');

    const handleEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      setEmail(event.target.value);
    };

    return (
      <Input
        label="Email"
        placeholder="Enter your email"
        variant={TYPE.SECOND}
        type="email"
        value={email}
        onChange={handleEmailChange}
        message={email ? '' : 'This field is required'}
        disabled={false}
      />
    );
  }
};

export const WithIcon: Story = {
  render: () => {
    const [password, setPassword] = useState('');

    const handlePasswordChange = (
      event: React.ChangeEvent<HTMLInputElement>
    ) => {
      setPassword(event.target.value);
    };

    return (
      <Input
        label="Password"
        placeholder="value..."
        variant={TYPE.PRIMARY}
        type="password"
        icon="/eye-show.svg"
        onChange={handlePasswordChange}
        onClick={() => alert('Icon clicked!')}
        value={password}
        message={password ? '' : 'Start typing...'}
      />
    );
  }
};

export const Textarea: Story = {
  render: () => {
    const [comments, setComments] = useState('');

    const handleCommentsChange = (
      event: React.ChangeEvent<HTMLTextAreaElement>
    ) => {
      setComments(event.target.value);
    };

    return (
      <Input
        label="Comments"
        placeholder="Write your comments..."
        variant={TYPE.THIRD}
        type="text"
        rows={4}
        value={comments}
        onChange={handleCommentsChange}
        message={comments.length > 20 ? 'Maximum 20 characters' : ''}
      />
    );
  }
};
