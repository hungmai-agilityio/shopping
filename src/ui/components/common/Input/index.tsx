import { forwardRef, InputHTMLAttributes } from 'react';
import { clsx } from 'clsx';
import Image from 'next/image';

// Constants
import { TYPE } from '@/constants';

interface InputProps
  extends InputHTMLAttributes<HTMLInputElement | HTMLTextAreaElement> {
  label?: string;
  variant?: TYPE.PRIMARY | TYPE.SECOND | TYPE.THIRD;
  htmlFor?: string;
  message?: string;
  icon?: string;
  type?: string;
  rows?: number;
  onClick?: () => void;
}

const Input = forwardRef<HTMLTextAreaElement | HTMLInputElement, InputProps>(
  (
    {
      label,
      value,
      variant,
      htmlFor,
      placeholder,
      name,
      onChange,
      onClick,
      message,
      icon,
      disabled,
      type = TYPE.TEXT,
      rows,
      onKeyDown
    },
    ref
  ) => {
    const typeClasses = {
      [TYPE.PRIMARY]: 'focus:border focus:border-green-300 h-12',
      [TYPE.SECOND]: 'border border-gray-400 h-12',
      [TYPE.THIRD]: `border border-gray-400 resize-none`
    };

    const commonClasses = clsx(
      'w-full mt-1.5 font-bold outline-none font-lg text-dark-gray px-2',
      { ' pr-10': icon }
    );
    const typeClass = typeClasses[variant || TYPE.PRIMARY];

    return (
      <div className="mb-3">
        {label && (
          <label
            htmlFor={htmlFor}
            className="font-bold leading-2-xs capitalize"
          >
            {label}
          </label>
        )}
        <div className="relative">
          {variant === TYPE.THIRD ? (
            <textarea
              ref={ref as React.Ref<HTMLTextAreaElement>}
              id={htmlFor}
              name={name}
              value={value}
              onChange={onChange}
              onKeyDown={onKeyDown}
              placeholder={placeholder}
              className={clsx('py-1 px-2', commonClasses, typeClass)}
              disabled={disabled}
              rows={rows}
            />
          ) : (
            <input
              ref={ref as React.Ref<HTMLInputElement>}
              id={htmlFor}
              name={name}
              value={value}
              onChange={onChange}
              onKeyDown={onKeyDown}
              type={type}
              placeholder={placeholder}
              className={clsx(commonClasses, typeClass)}
              disabled={disabled}
            />
          )}
          {icon && (
            <Image
              src={icon}
              alt="icon"
              onClick={onClick}
              width={24}
              height={24}
              className="absolute top-5 right-2 cursor-pointer"
            />
          )}
        </div>
        {message && (
          <span className="text-red-500 mr-1 visible text-xs">{message}</span>
        )}
      </div>
    );
  }
);

export default Input;
