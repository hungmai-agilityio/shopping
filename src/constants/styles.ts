import { SIZE, STATUS, TYPE } from "@/constants/enum";

// Button
export const buttonSizeClasses = {
  [SIZE.SMALL]: 'w-12 h-12 text-xs',
  [SIZE.MEDIUM]: 'md:text-lg md:w-btn-md w-28 text-xs',
  [SIZE.LARGE]: 'w-full text-base'
};

export const buttonTypeClasses = {
  [TYPE.PRIMARY]: 'bg-gradient-primary text-white font-bold',
  [TYPE.SECOND]: 'bg-white text-purple-500 font-bold',
  [TYPE.THIRD]:
    'w-10 p-2 bg-gray-200 rounded-full drop-shadow-lg text-lg focus:bg-second focus:text-white focus:font-bold'
};

// Input
export const inputTypeClasses = {
  [TYPE.PRIMARY]: 'focus:border focus:border-green-300 h-12',
  [TYPE.SECOND]: 'border border-gray-400 h-12',
  [TYPE.THIRD]: `border border-gray-400 resize-none`
};

// Table
export const cellSizeClasses = {
  [SIZE.SMALL]: 'lg:w-tb-sm',
  [SIZE.MEDIUM]: 'lg:w-tb-md',
  [SIZE.LARGE]: 'lg:w-tb-xl'
};

// Toast
export const toastStyles = {
  [STATUS.SUCCESS]: 'bg-green-500',
  [STATUS.ERROR]: 'bg-red-500',
  [STATUS.WARNING]: 'bg-yellow-500'
};