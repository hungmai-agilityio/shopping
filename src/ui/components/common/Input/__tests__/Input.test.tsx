import { render, screen } from '@testing-library/react';

// Constants
import { TYPE } from '@/constants';

// Components
import Input from '@/ui/components/common/Input';

// Test Data
const propsDefault = {
  label: 'Text Input',
  placeholder: 'Enter some text',
  value: 'test value',
  onChange: jest.fn(),
  message: 'This is an error message'
};

describe('Input', () => {
  it('renders PRIMARY type input', () => {
    render(<Input {...propsDefault} variant={TYPE.PRIMARY} />);

    const inputElement = screen.getByDisplayValue('test value');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveClass(
      'focus:border focus:border-green-300 h-12'
    );
  });

  it('renders SECOND type input', () => {
    render(<Input {...propsDefault} variant={TYPE.SECOND} />);

    const inputElement = screen.getByDisplayValue('test value');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveClass('border border-gray-400 h-12');
  });

  it('renders type textarea', () => {
    render(<Input {...propsDefault} variant={TYPE.THIRD} rows={6} />);
    const textareaElement = screen.getByDisplayValue(
      'test value'
    ) as HTMLTextAreaElement;

    expect(textareaElement).toBeInTheDocument();
    expect(textareaElement).toHaveClass('border border-gray-400 resize-none');
  });

  it('shows message when provided', () => {
    render(<Input {...propsDefault} message="This is an error message" />);

    const messageElement = screen.getByText('This is an error message');
    expect(messageElement).toBeInTheDocument();
  });

  it('should render match to snapshot', () => {
    const { container } = render(
      <Input {...propsDefault} variant={TYPE.PRIMARY} />
    );
    expect(container).toMatchSnapshot();
  });
});
