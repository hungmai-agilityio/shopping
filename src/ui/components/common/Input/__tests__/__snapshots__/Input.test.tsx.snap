// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render match to snapshot 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="font-bold leading-2-xs capitalize"
    >
      Text Input
    </label>
    <div
      class="relative"
    >
      <input
        class="w-full mt-1.5 font-bold outline-none font-lg text-dark-gray px-2 focus:border focus:border-green-300 h-12"
        placeholder="Enter some text"
        type="text"
        value="test value"
      />
    </div>
    <span
      class="text-red-500 mr-1 visible text-xs"
    >
      This is an error message
    </span>
  </div>
</div>
`;
